test/controllers/api_users/permissions_controller_test.rb

```diff
@@ -26,6 +26,69 @@ class ApiUsers::PermissionsControllerTest < ActionController::TestCase
       assert_not_authorised
     end
 
+    should "raise an exception if the user doesn't have access to the application" do
+      sign_in create(:admin_user)
+
+      application = create(:application)
+      api_user = create(:api_user)
+
+      assert_raises(ActiveRecord::RecordNotFound) do
+        get :edit, params: { api_user_id: api_user, application_id: application }
+      end
+    end
+
+    should "raise an exception if the application is retired" do
+      sign_in create(:superadmin_user)
+
+      application = create(:application, retired: true)
+      permission = create(:supported_permission, application:)
+      api_user = create(:api_user,
+                        with_signin_permissions_for: [application],
+                        with_permissions: { application => [permission.name] })
+      create(:access_token, application:, resource_owner_id: api_user.id)
+
+      assert_raises(ActiveRecord::RecordNotFound) do
+        get :edit, params: { api_user_id: api_user, application_id: application }
+      end
+    end
+
+    should "raise an exception if the user does not have an access token for the application" do
+      application = create(:application)
+      api_user = create(:api_user)
+
+      current_user = create(:superadmin_user)
+      sign_in current_user
+
+      assert_raises(ActiveRecord::RecordNotFound) do
+        get :edit, params: { api_user_id: api_user, application_id: application }
+      end
+    end
```

Comment:
In this test and the following one, should the user have the `signin` permission, to make sure we're just testing the missing access token, especially since they raise the same exception?